klog.go

```diff
@@ -823,19 +828,70 @@ func kvListFormat(b *bytes.Buffer, keysAndValues ...interface{}) {
 			v = missingValue
 		}
 		b.WriteByte(' ')
+		if k, ok := k.(string); ok {
+			// Avoid one allocation when the key is a string, which
+			// normally it should be.
+			b.WriteString(k)
+		} else {
+			b.WriteString(fmt.Sprintf("%s", k))
+		}
 
-		switch v.(type) {
-		case string, error:
-			b.WriteString(fmt.Sprintf("%s=%q", k, v))
+		// The type checks are sorted so that more frequently used ones
+		// come first because that is then faster in the common
+		// cases. In Kubernetes, ObjectRef (a Stringer) is more common
+		// than plain strings
+		// (https://github.com/kubernetes/kubernetes/pull/106594#issuecomment-975526235).
+		switch v := v.(type) {
+		case fmt.Stringer:
+			writeStringValue(b, true, v.String())
+		case string:
+			writeStringValue(b, true, v)
+		case error:
+			writeStringValue(b, true, v.Error())
 		case []byte:
-			b.WriteString(fmt.Sprintf("%s=%+q", k, v))
+			// We cannot use the simpler strconv.Quote here
+			// because it does not escape unicode characters, which is
+			// expected by one test!?
+			b.WriteByte('=')
+			b.WriteString(fmt.Sprintf("%+q", v))
```

Comment:
so multi-line is not supported for `[]byte`?

Reply:
Yes.

@serathius suggested that even strings should only be printed as multiline if the log call explicitly asks for it. I think that this would be error prone in practice because developers would have to think about whether the string might contain line breaks and how they want to handle that. We also would need some way of expressing that intend, which would be non-portable.

With `[]byte` I think the situation is a bit easier: if developers want plain text with potentially multiple lines, they can convert to string before logging. That seems fairly natural.

Reply:
I've rewritten the entire comment, trying also to capture why `%+q` might have been chosen.

Reply:
I still disagree with treating it as a string, but OK :)

Reply:
I chased the [original PR](https://github.com/kubernetes/klog/pull/237) which added `%+q` to [the issue](https://github.com/kubernetes/klog/issues/236) and from there to the [original complaint](https://github.com/kubernetes/kubernetes/issues/101714) which was about dumping a patch. Luckily someone [converted that value to a string](https://github.com/kubernetes/kubernetes/commit/3eadd1a9ead7a009a9abfbd603a5efd0560473cc), so that specific problem is solved.

I'm with @thockin, fixing that complaint by changing the formatting in the text output seems backwards. First, it doesn't solve the problem for JSON output. Treating the patch as string is also desired for JSON, but there the user gets a [base64 encoded string representation of the bytes](https://stackoverflow.com/a/34089818). Second, we now have the problem of whether such a byte array should support multi-line output.

I've reopened https://github.com/kubernetes/klog/issues/236 for further discussion.

However, that shouldn't block this PR because it doesn't change the handling of `[]byte` and isn't meant to.